### ISSUE TITLE
Supporting Distributed Frameworks in Pachyderm

### ISSUE THREAD

**Issue by dwhitena** [pachyderm/pachyderm#1623]
As we discussed in our 1.5 planning @gabrielgrant @JoeyZwicker, we need to:

- Determine if, when, and how we want to support distributed processing frameworks like Dask Distributed, Spark, and Distributed Tensorflow in Pachyderm, and
- Determine how we are going to guide users with respect to these sorts of tools (i.e., either direct them away from them, towards them within Pachyderm, or somewhere in between)

*Disclaimer* - This initial pass at discussion contained by very biased opinions, which we can abandon at any time.  I’m coming at this as a non- software engineer data scientist type of person, because that’s what I am.  

I would like to approach this problem in a few stages, which I think will allow us to zero in on what we need to do, if anything, that is Pachyderm specific (i.e., in terms of features, docs, or guides).  I think we should focus our discussion on the following questions and their implications:

1. What distributed types of processing are people using these types of frameworks for currently?
2. Which of these distributed types of processing are not handled well by Pachyderm currently?
3. Are there planned or possible features of Pachyderm (agnostic to any distributed frameworks) that would allow us to handle the distributed types of processing that aren’t currently handled by Pachyderm?
4. Are there scenarios when doing things the Pachyderm way (i.e., w/out invoking another distributed processing framework) isn’t enough or actually isn’t the best way, or are there scenarios when Pachyderm can’t do certain types of distributed processing?
5. In those scenarios how do we support distributed frameworks in Pachyderm?

However, before attacking those questions, I would like to express some of my opinions about the “secret sauce” that Pachyderm provides.  These are things that, in my opinion, differentiate Pachyderm in a good way and that we don’t want to pollute by allowing users to do “anything” in Pachyderm (e.g., like running a whole Hadoop cluster that is utilized in pipelines).

## The Pachyderm (non-)secret sauce

One of the things that really, really excites me about Pachyderm is the fact that we are, in many cases, allowing data scientists to avoid complicating their code and implementations because of scaling issues.  For example, prior to Pachyderm, a typical data science situation scenario went like this:

- Data scientist implements cool stuff using some python framework
- Data scientist wants to put this into production
- Data Engineer says “are you crazy, you’re gonna run out of memory and destroy our lives.”
- Data scientist realizes that Data Engineer is correct, is sad and either complicates their code with some other framework or has to pass their solution off to engineers to re-implement in some other language or framework
- Everyone is unhappy because this is not efficient and, at the end of the day, ownership and understanding about the solution is very convoluted.

Whereas, with Pachyderm, everyone is happy because:

- Data scientist implements cool stuff using some python framework
- Data scientist wants to put this into production
- Data scientist puts it into production by putting it in a Pachyderm pipeline
- Everyone is happy and gets off work early.

Now, there are definitely cases where this sort of data science nervana isn’t totally possible with Pachyderm, but for the most part, I would describe this as part of Pachyderm “secret sauce.”  

What we shouldn’t do (again in my opinion) is encourage and make it easy for data people to do all the things they are currently doing in the same way they are currently doing them in Pachyderm.  This kind of defeats the purpose.  We want people to have a new, better experience with Pachyderm and not have to resort to code re-writing and distributed frameworks whenever there is a simple and sustainable “Pachyderm” way of doing it.

I think what we need to determine is if there are scenarios when doing things the Pachyderm way isn’t enough or actually isn’t the best way, or if there are scenarios when Pachyderm can’t do certain types of distributed processing.  These are the types of scenarios, I think, that would warrant a new, potentially complicated, feature(s) of Pachyderm to enable distributed processing framework within the already distributed processing of Pachyderm.

## What distributed types of processing are people using these types of frameworks for currently?

Popular distributed processing frameworks that have come up in our discussions are Dask Distributed, Spark, and Distributed Tensorflow, so let’s take these frameworks as examples. Generally, I think that we can fit the types of distributed processing or optimizations offered by these frameworks into the following categories:

- data locality (e.g., Spark’s caching or Dask optimizations)
- map operations 
- reduce/gather operations
- distributed joins (e.g., Dask/Spark dataframe merges)
- shuffle operations 
- distributed ad hoc queries (e.g., Spark SQL, Dask Distributed dataframes)
- online, streaming operations (e.g., Spark streaming, Storm)
- distribute graph computations (e.g., Spark GraphX)
- asynchronous/synchronous model training (e.g., TensorFlow tf.train.SyncReplicasOptimizer)

## Which of these distributed types of processing are not handled well by Pachyderm currently?

In my understanding, distributed types of processing that are handled well by Pachyderm currently are:

- map operations (via `/*` globs)
- shuffle operations (via directory organization, soon to be better)
- online, streaming operations (via frequent commits and automatically syncing results)

Distributed types of processing that are possible, but maybe not convenient, in Pachyderm include:

- reduce/gather operations (via `/` globs or succesive `/*` pipelines) - this is slightly inconvenient because doing a partitioned join would involve manually defining multiple stages which do successive portions of the reduce.
- distributed joins (via `/*` globs and cross-product distribution of datums) - this is slightly inconvenient because you need to still split into multiple files while possibly managing a header.

Distributed types of processing that are, to my knowledge, not entirely possible in Pachyderm:

- data locality - because PFS data is necessarily separated from compute.  That being said, we have already discussed caching improvements (which are actually similar to other systems), so I don’t think we need to discuss this topic much more here.
- distributed ad hoc queries - I think we are exploring various ways of enabling SQL within pachyderm, but even with that, I’m not sure how we would actually do distributed queries similar to those of Spark SQL.
- distribute graph computations - Maybe this is possible.  Maybe it is not.  Either way, I don’t think it’s a big enough use case at the moment for us to consider.
- asynchronous/synchronous model training - This is a complicated subject, and I definitely don’t think we could cover this right now.  The best we could do is throw certain types of nodes at the training (e.g., GPUs).  It may be possible to implement some type of pipeline that does this, but it’s far from clear to me how we would do that.

## Are there planned or possible features of Pachyderm (agnostic to any distributed frameworks) that would allow us to handle the distributed types of processing that aren’t currently handled by Pachyderm?

As mentioned above, I think the only planned features of Pachyderm that could help with data locality and model training are:

- Caching strategies
- Node selector functionality (e.g., for GPUs)

For model training, the second one will actually go pretty far, because most orgs aren’t even thinking about distributed model training.  They just need to train and get a model into production.  That is the biggest blocker.  Also, even though there is a lot of hype around TensorFlow and Deep Learning, most business use cases don’t require deep convolutional neural networks, and the majority of practical machine learning doesn’t require distributed training.

That being said, distributed deep learning, like Tensorflow, is definitely the hype and gets a lot of attention.  Thus, it might be good to think about that as it picks up steam.

Regarding “possible” Pachyderm features, maybe @JoeyZwicker can comment of certain plans around database and query support in Pachyderm?  I don’t think this is super clear yet, at least to me.

## Are there scenarios when doing things the Pachyderm way (i.e., w/out invoking another distributed processing framework) isn’t enough or actually isn’t the best way, or are there scenarios when Pachyderm can’t do certain types of distributed processing?

I think given the above answers, where we might want to focus our attention on either enabling better distributed computation in Pachyderm or enabling other distributed frameworks within Pachyderm are:

- distributed ad hoc queries 
- asynchronous/synchronous model training 
- (maybe joins)

These seem to be the most relevant types of distributed processing that people just can’t do in a “Pachyderm” way, and where it isn’t clear that doing things via globs and pipeline stages are going to help up.

In these cases, maybe it is worth figuring out how to enable things like Distributed Dask or Tensorflow within Pachyderm.  This makes me excited but also a little scared.  I think we should be really careful about how we encourage users to do these things, and only resort to using these types of frameworks for operations like ad hoc queries or model training.  In other cases, they should keep things simple and do it the Pachyderm way.  This will save them huge headaches.

## In those scenarios how do we support distributed frameworks in Pachyderm?

In both the distributed Dask and distributed Tensorflow cases (ignoring Spark for now, because I don’t think it really makes that many people happy), we would need to allow a Pachyderm pipeline stage to be executed collaboratively by multiple “workers” (putting workers in quotes here, because I’m not sure it really corresponds to our current idea of workers). 

To me, it seems like there are a few options to enable this collaborative execution:

1. have a single Pachyderm worker spin up a distributed Dask/Tensorflow cluster within k8s and offload its work to that cluster, gathering results.
2. have a pre-existing Dask/Tensorflow cluster spun up in k8s that could be specified in the pipeline spec, similar to selecting a certain type of node to run a pipeline.
3. have a Dask/Tensorflow cluster spun up as a service within Pachyderm that is accessible to all pipeline stages.
4. create some type of Docker abomination that has a docker in docker deployment of a distributed framework.

To me, options 1 and 2 seem like the best solutions, or at least the ones that might require the least amount of friction as is.  My recommendation is trying the following:

1. Spin up TensorFlow or some other framework within k8s.
2. Without modification, try utilizing that within a Pachyderm pipeline stage (e.g., via a certain exposed port).  This isn’t a great solution, but it would at least give us a quick way to answer people for the time being with a solid example.
3. As we are exploring #2, we should determine what it would take to provide some type of selector functionality within the pipeline spec to allow uses to offload a certain workload from a worker on to an existing distributed framework deployed within k8s.
4. We implement the selector functionality or determine that it’s not the right solution.

**Comment by gabrielgrant** [comment]
@dwhitena  when you say "distributed ad hoc queries" and/or "online, streaming operations" are you talking about live, exploratory analysis?

because we are more production-oriented than many (most?) other tools, doing quick, exploratory (possibly throw-away) analysis has more friction than i'd like right now. hoping this may be something that can at least partially be addressed by integrating jupyter with the UI, but in the simplest implementation (using something like the 1.3 service model), we still don't really solve the problem of wanting a sandbox backed by distributed computation, since the service would still be running on a single instance

**Comment by dwhitena** [comment]
@gabrielgrant By online/stream operations I meant distributed real-time processing, where all of you data does not fit into memory either by choice or necessity.  By ad hoc queries, I meant things like Spark SQL or distributed dataframe operations, so probably "ad hoc" wasn't good wording.  Basically, I meant the ability to execute arbitrary queries, select, subset, etc. operations without previously arranging the data in such a way to facilitate specific queries.

I think Jupyter, etc. are a slightly different conversation, although maybe related in some ways as you have noted (Spark notebooks, zeppelin, Jupyter + Spark, etc.).

**Comment by JoeyZwicker** [comment]
There isn't a clear DB-style or query support path for Pachyderm yet. As you guys mention in the ad-hoc discussion above, getting information/queries that your data wasn't specifically arranged to support well isn't really feasible right now. We occasionally have users who want to "export pachyderm metadata" in some fashion to a DB such as elastic search to make these types of queries possible.

**Comment by gabrielgrant** [comment]
I'm not sure I totally understand what kind of operations are enabled by exporting pach metadata to an external DB... is this to somehow allow ad-hoc querying of their data? or to look into their repos'/pipelines' history or something?

**Comment by JoeyZwicker** [comment]
I dont think this was very well thought out, but it was something being asked a few weeks ago. basically they wanted to be able to query for file names and figure out which commits and repos contained a file or, for example, which jobs have processed a given file. They wanted a way to use a query language to do that instead of pachctl and shell.

**Comment by gabrielgrant** [comment]
This is maybe something to discuss again for 1.7 now that services are making a comeback (see: #2270 )

**Comment by westurner** [comment]
For online queries from various datastores: https://github.com/blaze/blaze translates Python code to {SQL,}

Dask Distributed is mentioned above. Dask-ml is fairly new: https://dask-ml.readthedocs.io/en/latest/

From https://dask-ml.readthedocs.io/en/latest/joblib.html :

> For large arguments that are used by multiple tasks, it may be more efficient to pre-scatter the data to every worker, rather than serializing it once for every task. This can be done using the `scatter` keyword argument, which takes an iterable of objects to send to each worker.

```python
# Serialize the training data only once to each worker
with parallel_backend('dask.distributed', scheduler_host='localhost:8786',
                      scatter=[digits.data, digits.target]):
    search.fit(digits.data, digits.target)
```
dask-ml is now installed in the kaggle/docker-python image:
https://github.com/Kaggle/docker-python/blob/master/Dockerfile

**Comment by westurner** [comment]
What are the advantages of pachyderm over e.g. dask-ml w/ joblib? How much k8s and provenance tracking does it do for me? W3C PROV?

I notice that you have used protobuf; whereas parquet is built on thrift. 
https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift

**Comment by westurner** [comment]
Would Apache Arrow make this data handoff potentially more efficient (as it does between Spark and Pandas dataframes)?

**Comment by nirmalsinghania2008** [comment]
Any updates on this?

**Comment by JoeyZwicker** [comment]
@nirmalsinghania2008 Actually yes, we're actively working on this right now and there have already been a number of features added that make this doable in Pachyderm already. S3gateway (#3400) and #3099 and were working on making this really powerful with #3513

**Comment by nirmalsinghania2008** [comment]
Nice, will experiment with this. what is the current preferred way to run distributed tensorflow with pachyderm?

**Comment by gabrielgrant** [comment]
@nirmalsinghania2008 sorry for being so slow to get back to this. I've opened a separate issue specifically related to the state of Kubeflow-Pachyderm integration here: #3776
